### PR TITLE
feat(queryRules): add ais-query-rule-custom-data widget

### PR DESCRIPTION
**Pull request by Haroenv**
This adds the `ais-query-rule-custom-data` widget based on the InstantSearch.js `queryRules` connector.

### CODE REVIEW

File: examples/storybook/src/stories/Hits.stories.ts [hunk 1]
```diff
@@ -8,4 +8,19 @@ storiesOf('Hits', module)
     component: wrapWithHits({
       template: '<ais-hits></ais-hits>',
     }),
+  }))
+  .add('customized hits', () => ({
```
Comment (Haroenv) [inline examples/storybook/src/stories/Hits.stories.ts]: just an unrelated small fix I used as reference making my stories

File: examples/storybook/src/stories/QueryRuleCustomData.stories.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                [alt]="item.title"
+                [style.width]="'100%'"
+              />
+            </a>
+          </div>
+        </ng-template>
+      </ais-query-rule-custom-data>`,
+      methods: {
+        transformItems: items => {
+          console.log('hi', items);
```
Comment (tkrugg) [inline examples/storybook/src/stories/QueryRuleCustomData.stories.ts]: oops 😄

Comment (Haroenv) [reply examples/storybook/src/stories/QueryRuleCustomData.stories.ts]:
```suggestion
```

File: examples/storybook/src/stories/QueryRuleCustomData.stories.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    component: wrapWithHits({
+      ...moviesConfig,
+      template: `
+      <p>
+        Type <q>not a banner</q> and nothing will appear.
+      </p>
+      <ais-query-rule-custom-data [transformItems]="transformItems">
+      </ais-query-rule-custom-data>`,
+      methods: {
+        transformItems: (items) =>
```
Comment (Haroenv) [inline examples/storybook/src/stories/QueryRuleCustomData.stories.ts]:
```suggestion
        transformItems: items =>
```

File: src/query-rule-custom-data/query-rule-custom-data.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          <pre>{{ item | json }}</pre>
+        </div>
+      </div>
+    </div>
+  `,
+})
+export class NgAisQueryRuleCustomData extends BaseWidget {
+  @ContentChild(TemplateRef) public template: any;
+
+  @Input() public transformItems?: <U extends {}>(items: {}[]) => U[];
```
Comment (tkrugg) [inline src/query-rule-custom-data/query-rule-custom-data.ts]: does actually `U` need to extend the original type for item? I'm thinking maybe having: `<U>(items: {}[]) => U[];`

File: src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { createRenderer } from '../../../helpers/test-renderer';
+import { NgAisQueryRuleCustomData } from '../query-rule-custom-data';
+
+const render = createRenderer({
+  defaultState: {
+    items: [{ banner: 'lol yes it is a banner' }],
+  },
+  template: '<ais-query-rule-custom-data></ais-query-rule-custom-data>',
```
Comment (tkrugg) [inline src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts]: We could add a test for the case using `context` to cover that use case too.

Comment (Haroenv) [reply src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts]: I'm not following what you mean. Do you want a test with both widgets?

Comment (tkrugg) [reply src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts]: I just mean write a test with the following template: 
```html
<ais-query-rule-custom-data>
   <ng-template>
    ...
   </ng-template>
</ais-query-rule-custom-data>
```
I think it would make sense since it's a different feature, and if a contributor breaks it inadvertently, a test would fire up an alarm.

Comment (Haroenv) [reply src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts]: ah you mean "items"

Comment (tkrugg) [reply src/query-rule-custom-data/__tests__/query-rule-custom-data.spec.ts]: yes, or just ng-template 😄

File: src/query-rule-custom-data/query-rule-custom-data.ts (hunk 4 above)
Comment (francoischalifour) [reply src/query-rule-custom-data/query-rule-custom-data.ts]: `any` is the most correct type that we can use as long as the component is not generic.

See algolia/instantsearch.js#3670.

**Comment by francoischalifour** [comment]
I added the InstantSearch.js 3.3.0 dependency and added tests.

File: examples/storybook/src/stories/QueryRuleCustomData.stories.ts [hunk 6]
```diff
@@ -99,33 +112,4 @@ storiesOf('QueryRuleCustomData', module)
...
-        transformItems: items => [items[0]],
-      },
-    }),
-  }))
-  .add('keeping only banners with transform', () => ({
-    component: wrapWithHits({
-      ...moviesConfig,
-      template: `
-      <p>
-        Type <q>not a banner</q> and nothing will appear.
```
Comment (Haroenv) [inline examples/storybook/src/stories/QueryRuleCustomData.stories.ts]: I thought this was a good use case for transformItems

Comment (francoischalifour) [reply examples/storybook/src/stories/QueryRuleCustomData.stories.ts]: Yes, not exactly sure why I removed them. I added them back!

File: src/index.ts [hunk 7]
```diff
@@ -45,6 +45,8 @@ import { NgAisPanelModule } from './panel/panel.module';
 export { NgAisPanelModule };
 import { NgAisConfigureModule } from './configure/configure.module';
 export { NgAisConfigureModule };
+import { NgAisQueryRuleCustomDataModule } from './query-rule-custom-data/query-rule-custom-data.module';
```
[the remark this answers is not in the document]
Comment (francoischalifour) [reply src/index.ts]: I just double checked and this is how Prettier formats it.
